### PR TITLE
Reloadable config

### PR DESCRIPTION
**Pull request by ncarroll-grindr**
This PR adds:
- ReloadableConfig, which will reload the config based on an added Watchable
- Watchable interface, which provides a callback used to reload the config
- ConsulWatcher, which reloads the config when a consul KVCache detects an update
- FileWatcher, which reloads the config when there are changes to a directory

### CODE REVIEW

File: hoplite-watch/src/main/kotlin/com/sksamuel/hoplite/watch/ReloadableConfig.kt [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package com.sksamuel.hoplite.watch
+
+import com.sksamuel.hoplite.ConfigLoader
+import com.sksamuel.hoplite.fp.getOrElse
+import java.util.concurrent.atomic.AtomicReference
+import kotlin.reflect.KClass
+
+class ReloadableConfig<A : Any>(private val configLoader: ConfigLoader, private val clazz: KClass<A>) {
+  private val config = AtomicReference(configLoader.loadConfig(clazz, emptyList()).getOrElse { null })
```
Comment (ncarroll-grindr) [inline hoplite-watch/src/main/kotlin/com/sksamuel/hoplite/watch/ReloadableConfig.kt:9]: I figured we probably didn't want to throw here, so the config is optional.

File: hoplite-watch/src/main/kotlin/com/sksamuel/hoplite/watch/watchers/consul.kt [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+import com.orbitz.consul.cache.KVCache
+import com.sksamuel.hoplite.watch.Watchable
+
+class ConsulWatcher(private val kvCache: KVCache) : Watchable {
+  override fun watch(callback: () -> Unit) {
+    kvCache.addListener {
+      callback()
+    }
+    kvCache.start()
```
Comment (ncarroll-grindr) [inline hoplite-watch/src/main/kotlin/com/sksamuel/hoplite/watch/watchers/consul.kt:11]: We could also accept a string key + client and handle the kvCache creation in this method, but this gives callers a little more flexibility.

File: hoplite-watch/src/test/kotlin/com/sksamuel/hoplite/watch/WatcherTest.kt [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+      .addSource(PropertySource.file(tmpFile))
+      .build()
+
+    val reloadableConfig = ReloadableConfig(configLoader, TestConfig::class)
+      .addWatcher(FileWatcher(tmpFile.parent))
+
+    val config = reloadableConfig.getLatest()
+    config?.foo shouldBe "bar"
+
+    tmpFile.writeText("""{"foo": "baz"}""")
```
Comment (ncarroll-grindr) [inline hoplite-watch/src/test/kotlin/com/sksamuel/hoplite/watch/WatcherTest.kt:81]: Still fixing this test. I can get this to work with a Thread.sleep(...) before the file change - maybe this is firing before the FileWatcher starts polling?

File: buildSrc/src/main/kotlin/Libs.kt [hunk 4]
```diff
@@ -67,4 +68,8 @@ object Libs {
   object Vavr {
     const val kotlin = "io.vavr:vavr-kotlin:0.10.2"
   }
+
+  object EmbeddedConsul {
+    const val consul = "com.pszymczyk.consul:embedded-consul:2.2.1"
```
Comment (sksamuel) [inline buildSrc/src/main/kotlin/Libs.kt:73]: Oooo we should add integration with kotest.

File: hoplite-watch/src/main/kotlin/com/sksamuel/hoplite/watch/watchers/consul.kt (hunk 2 above)
Comment (sksamuel) [reply hoplite-watch/src/main/kotlin/com/sksamuel/hoplite/watch/watchers/consul.kt:11]: Should we be starting the cache?

File: hoplite-watch/src/main/kotlin/com/sksamuel/hoplite/watch/watchers/file.kt [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    pathToWatch.register(
+      watchService,
+      StandardWatchEventKinds.ENTRY_CREATE,
+      StandardWatchEventKinds.ENTRY_MODIFY,
+      StandardWatchEventKinds.ENTRY_DELETE
+    )
+
+    Executors.newSingleThreadExecutor().submit {
+      while (true) {
+        val watchKey = watchService.take()
```
Comment (sksamuel) [inline hoplite-watch/src/main/kotlin/com/sksamuel/hoplite/watch/watchers/file.kt:30]: Does this block. If not we will want a sleep in the loop

File: hoplite-watch/src/test/kotlin/com/sksamuel/hoplite/watch/WatcherTest.kt (hunk 3 above)
Comment (sksamuel) [reply hoplite-watch/src/test/kotlin/com/sksamuel/hoplite/watch/WatcherTest.kt:81]: Could move the write into the eventually and randomise some character.

File: settings.gradle.kts [hunk 6]
```diff
@@ -25,3 +25,5 @@ include("hoplite-json")
 include("hoplite-toml")
 include("hoplite-yaml")
 include("hoplite-vavr")
+include("hoplite-watch")
+include("test")
```
Comment (sksamuel) [inline settings.gradle.kts:29]: What is the include test for

**Comment by sksamuel** [comment]
Couple of improvements:
- put the consul watcher in a package `hoplite-watcher-consul` so people don't have to bring that in just to use watcher.
- Add a callback to the reloadable which is invoked if the reloadConfig throws.

File: hoplite-watch/src/main/kotlin/com/sksamuel/hoplite/watch/watchers/consul.kt (hunk 2 above)
Comment (ncarroll-grindr) [reply hoplite-watch/src/main/kotlin/com/sksamuel/hoplite/watch/watchers/consul.kt:11]: Good question, I think it depends:
- not starting it puts the impetus on the caller to start, and makes configuration a bit less simple
- starting could cause problems if the consumer stops it and we're not longer receiving events

File: hoplite-watch/src/main/kotlin/com/sksamuel/hoplite/watch/watchers/file.kt (hunk 5 above)
Comment (ncarroll-grindr) [reply hoplite-watch/src/main/kotlin/com/sksamuel/hoplite/watch/watchers/file.kt:30]: The docs say `Retrieves and removes next watch key, waiting if none are yet present.` which makes me think it does block.

File: hoplite-watch/src/test/kotlin/com/sksamuel/hoplite/watch/WatcherTest.kt (hunk 3 above)
Comment (ncarroll-grindr) [reply hoplite-watch/src/test/kotlin/com/sksamuel/hoplite/watch/WatcherTest.kt:81]: I think the issue there is that the events don't fire immediately, so the config may never match the most recently written value.

File: settings.gradle.kts (hunk 6 above)
Comment (ncarroll-grindr) [reply settings.gradle.kts]: Removed!

**Comment by ncarroll-grindr** [comment]
Moved to a separate module, and added a callback to the reloadable.

File: hoplite-watch/src/test/kotlin/com/sksamuel/hoplite/watch/WatcherTest.kt (hunk 3 above)
Comment (sksamuel) [reply hoplite-watch/src/test/kotlin/com/sksamuel/hoplite/watch/WatcherTest.kt:80]: Ok that's a problem. Go with a delay(4000) then ?

File: hoplite-watch/src/main/kotlin/com/sksamuel/hoplite/watch/watchers/file.kt (hunk 5 above)
Comment (sksamuel) [reply hoplite-watch/src/main/kotlin/com/sksamuel/hoplite/watch/watchers/file.kt:30]: Sounds like it.
Put everything inside the while (true) in a try catch block. .take() can throw if an interrupt fires and we don't want the while loop to exit if so.

**Comment by sksamuel** [comment]
Ok I think fix the flakey test, add in the try block inside the while true and we can merge.

**Pull request merged**